src/auto-router.ts

```diff
@@ -215,127 +255,110 @@ export class AutoRouter {
-    if (!priceOracleAddress) throw Error("No default PhuturePriceOracle found for chain");
+        const { minBuyAmount } = await this.zeroExAggregator.allowanceHolderQuote({
+          ...zeroExOptions,
+          chainId,
+          sellToken: routerSellTokenAddress,
+          buyToken: asset,
+          sellAmount: amount.toString(),
+          taker: this.indexRouter.contract.address,
+        });
 
-    const priceOracle = PhuturePriceOracle__factory.connect(
-      priceOracleAddress,
-      this.indexRouter.signer,
+        return {
+          asset,
+          minBuyAmount,
+        };
+      }),
     );
 
+    const priceOracleAddress = defaultPhuturePriceOracleAddress[chainId];
+    if (!priceOracleAddress) throw new Error("No default PhuturePriceOracle found for chain");
+    const priceOracle = PhuturePriceOracle__factory.connect(priceOracleAddress, this.indexRouter.signer);
+
     const buyAmountsInBase = await Promise.all(
-      buyAmounts.map(async ({ asset, buyAssetMinAmount }, amountIndex) => {
+      buyAmounts.map(async ({ asset, minBuyAmount }, amountIndex) => {
         const price = await priceOracle.callStatic.refreshedAssetPerBaseInUQ(asset);
         return {
           asset,
-          buyAmount: BigNumber.from(buyAssetMinAmount)
-            .mul(BigNumber.from(2).pow(112))
-            .mul(255)
-            .div(price.mul(amounts[amountIndex]!.weight)),
+          buyAmount: BigNumber.from(minBuyAmount)
+            .mul(UQ112)
+            .mul(MAX_WIDTH)
+            .div(price.mul(indexAnatomy[amountIndex].weight)),
         };
       }),
     );
 
-    const minAmount = buyAmountsInBase.reduce((min, curr) =>
-      min.buyAmount.lte(curr.buyAmount) ? min : curr,
+    const minAmount = buyAmountsInBase.reduce((min, curr) => (min.buyAmount.lte(curr.buyAmount) ? min : curr));
+
+    const scaledSellAmounts = initialBuyAmounts.map(({ amount }, i) =>
+      amount.mul(minAmount.buyAmount).div(buyAmountsInBase[i].buyAmount),
     );
 
-    const scaledSellAmounts = Object.values(
-      amounts.map(({ asset, weight }) => ({
-        asset,
-        amount: BigNumber.from(sellAmount).mul(weight).div(255),
-        weight,
-      })),
-    ).map(({ amount }, i) => amount.mul(minAmount.buyAmount).div(buyAmountsInBase[i].buyAmount));
+    const totalSellAmount = scaledSellAmounts.reduce((sum, curr) => sum.add(curr), BigNumber.from(0)).toString();
 
-    const quotes = await Promise.all(
-      amounts.map(async ({ asset }, i) => {
-        const scaledSellAmount = scaledSellAmounts[i] as BigNumber;
-        if (asset === sellToken || scaledSellAmount.isZero())
+    const finalQuotes = await Promise.all(
+      indexAnatomy.map(async ({ asset }, i) => {
+        const scaledAmount = scaledSellAmounts[i] as BigNumber;
+
+        // If asset is sell token or amount is zero, no swap needed
+        if (isAddressEqual(asset as Address, sellToken as Address) || scaledAmount.isZero()) {
```

Comment:
are you sure you need to use a quote with zero amount? Maybe you should skip it?